Mathlib/Topology/Order/UpperSetTopology.lean

```diff
@@ -0,0 +1,283 @@
+/-
+Copyright (c) 2023 Christopher Hoskin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christopher Hoskin
+-/
+import Mathlib.Topology.Homeomorph
+import Mathlib.Topology.Order.Lattice
+import Mathlib.Order.Hom.CompleteLattice
+import Mathlib.Tactic.TFAE
+import Mathlib.Topology.Order.LowerTopology
+
+/-!
+# UpperSet topology
+
+This file introduces the upper set topology on a preorder as the topology where the open sets are
+the upper sets.
+
+In general the upper set topology does not coincide with the dual of the lower topology.
+
+## Main statements
+
+- `UpperSetTopology.IsOpen_sInter` - the intersection of any set of open sets is open
+- `UpperSetTopology.IsOpen_iInter` - the intersection of any indexed collection of open sets is open
+- `UpperSetTopology.isClosed_iff_isLower` - a set is closed if and only if it is a Lower set
+- `UpperSetTopology.closure_eq_lowerClosure` - topological closure coincides with lower closure
+- `UpperSetTopology.Monotone_tfae` - the continuous functions are characterised as the monotone
+  functions
+- `UpperSetTopology.Monotone_to_LowerTopology_Dual_Continuous` - a `Monotone` map from a `Preorder`
+  with the `UpperSetTopology` to the dual of a `Preorder` with the `LowerTopology` is `Continuous`
+
+## Implementation notes
+
+A type synonym `WithUpperSetTopology` is introduced and for a preorder `α`, `WithUpperSetTopology α`
+is made an instance of `TopologicalSpace` by the topology where the upper sets are the open sets.
+
+We define a mixin class `UpperSetTopology` for the class of types which are both a preorder and a
+topology and where the open sets are the upper sets. It is shown that `WithUpperSetTopology α` is an
+instance of `UpperSetTopology`.
+
+## Motivation
+
+An Alexandrov topology is a topology where the intersection of any collection of open sets is open.
+The `UpperSetTopology` is an Alexandrov topology, and, given any Alexandrov topology, a `Preorder`
+may be defined on the underlying set such that the `UpperSetTopology` of the `Preorder` coincides
+with the original topology.
+
+Furthermore, the `UpperSetTopology` is used in the construction of the Scott Topology.
+
+## Tags
+
+upper set topology, preorder, Alexandrov
+-/
+
+
+variable (α β : Type _)
+
+section preorder
+
+variable [Preorder α]
+
+/--
+The set of upper sets forms a topology. In general the upper set topology does not coincide with the
+dual of the lower topology.
+-/
+def upperSetTopology' : TopologicalSpace α :=
```

Comment:
This may bite us back at some point, but what do you think of just calling that `upperSetTopology` (with a small `u`) and using the case variation to distinguish it from `UpperSetTopology`. If you think that will be annoying I have no problem with keeping the current name.

Reply:
I find having `upperSetTopology'`, `UpperSetTopology` and `WithUpperSetTopology` a bit confusing. Originally I tried not naming `upperSetTopology'` at all, but found I seemed to keep needing to refer to it.

Reply:
I think it makes sense to have all three (see my comments on Zulip), as long as you are a bit careful about how you build up the API. The current naming is fine, so do whatever you prefer.